PYIC-5669: New Screen: What does the user want to change

sammyiyke opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
<!-- Include the Jira ticket number in square brackets as prefix, eg `[P4-XXXX] PR Title` -->

## Proposed changes
[PYIC-5669](https://govukverify.atlassian.net/browse/PYIC-5669): New Screen: What does the user want to change

Creation of a new page titled "Update your details" as part of the "Continuation of Identity" initiative for IPV core 

### What changed

- Adds new "Update your details" page/screen for confirming and selecting user detail changes as part of the "Continuation of Identity" initiative
- Adds support for new data format (`formattedName`) returned to user details in tests

<!-- Describe the changes in detail - the "what"-->

### Why did it change

As part of the COI user journey we need to allow users choose what details they would like to change.

This feature is critical for maintaining accurate and up-to-date user information in accordance with GPG45 specifications.

<!-- Describe the reason these changes were made - the "why" -->

### Issue tracking
<!-- List any related Jira tickets or GitHub issues -->
<!-- List any related ADRs or RFCs -->
<!-- Delete/copy as appropriate -->

- [PYIC-5669](https://govukverify.atlassian.net/browse/PYIC-5669)

<img width="1512" alt="Screenshot 2024-04-16 at 11 11 38" src="https://github.com/govuk-one-login/ipv-core-front/assets/11758541/acd871f5-75f2-4396-8b80-58cafe6c2658">

<img width="347" alt="Screenshot 2024-04-16 at 11 12 00" src="https://github.com/govuk-one-login/ipv-core-front/assets/11758541/ebc5dc3d-7f32-42a3-8296-f7c90c2611af">




[PYIC-5669]: https://govukverify.atlassian.net/browse/PYIC-5669?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[PYIC-5669]: https://govukverify.atlassian.net/browse/PYIC-5669?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

sammyiyke committed:
PYIC-5669 - Adds new screen and text - pyi-update-details

- Add new page/screen for confirming and selecting user detail changes
- Add support for checkboxes on new page
- **Missing** -- Currently missing the checkbox hint color as in the UCD

sammyiyke committed:
PYIC-5669 - Adds support for new data format to user details

sammyiyke committed:
PYIC-5669 - Fixes failing test

Adds support for formattedName to tests

sammyiyke committed:
PYIC-5669 - Add new ACs updates

- Updates screen to reflect new changes in the ticket's acceptance criteria
- Adds error handling

RogVanGDS commented on src/app/shared/reuseHelper.js:
```diff
@@ -52,6 +56,7 @@ module.exports = {
   generateUserDetails: (userDetailsResponse, i18n) => {
     return {
       name: userDetailsResponse.data?.name,
+      formattedName: userDetailsResponse.data?.formattedName,
```
Wish we had typescript to make this a bit more obvious what was going on 😢 

Could we rename "name" to "fullName" at least to help distinguish between the two?

sammyiyke replied:
My reasoning here is to leave as much as possible unchanged since a new field is being added. I did rename it to fullName earlier but change it cause I felt the `name` keeps some consistency with what is sent from core-back too.

Just my thoughts.. But I am happy to be wrong..

RogVanGDS commented on src/locales/en/translation.json:
```diff
@@ -407,6 +407,33 @@
+          "hint": "Select all that apply.",
+          "errorMessage": "Select at least one option",
+          "checkbox1": "Given names",
+          "checkbox1Hint": "This is your first name and any middle names",
```
Not sure we need the `"any"` looking at the designs

RogVanGDS commented on src/views/ipv/page/pyi-update-details.njk:
```diff
@@ -0,0 +1,109 @@
+              fieldset: {
+                legend: {
+                  text: 'pages.pyiUpdateDetails.content.checkboxes.heading' | translate,
+                  isPageHeading: true,
```
I think we don't want `isPageHeading: true`, the design docs suggest that it should be set to true if the legend also acts as the heading for the page.

I would assume that this isn't the case given we have an h1 header above it in the page?

RogVanGDS replied:
(We should only be using this if we want to avoid the screenreader from repeating the same heading twice, if the heading of the page is the same as the radio button header)

sammyiyke replied:
Oh, didn't know that. Good to know 👍

RogVanGDS commented on src/views/ipv/page/pyi-update-details.njk:
```diff
@@ -0,0 +1,109 @@
+                  divider: 'pages.pyiUpdateDetails.content.checkboxes.checkbox5Divisor' | translate
+                },
+                {
+                  value: "none",
```
Slightly nit-picky I know, but could we change the value to be something like `back` / `cancel` instead of `none`? 

Personally `none` suggests that we are intentionally leaving it empty, instead of it being a valid journey event

sammyiyke committed:
PYIC-5669 - PR related fixes

sammyiyke committed:
PYIC-5669 - Change - formattedName to nameParts

Changes formattedName to nameParts to match changes to data sent from core-back

Changes format for GivenNames and FamilyName

Fixes rendering for new name format in template and tests

sammyiyke committed:
PYIC-5669 - PR comments related

Changed value: "none" to "cancel" to indicate a valid Journey event

Joe-Edwards-GDS commented on src/views/ipv/page/pyi-update-details.njk:
```diff
@@ -0,0 +1,122 @@
+      {% if namePart.type == "FamilyName" %}
+        {% set familyNames = (familyNames.push(namePart.value), familyNames) %}
+      {% endif %}
+    {% endfor %}
```
I'd probably slightly prefer this logic to be done in `reuseHelper.js` - it'll be easier than here, and it'll be reusable if we need to display the given/family name on another page.

Not critical - we could wait until we know we need to reuse it.

sammyiyke replied:
Makes sense. Didn't think of it.

Joe-Edwards-GDS commented on src/views/ipv/page/pyi-update-details.njk:
Can we update this path to be simply `update-details` - I think going forward we're not using the `pyi` prefix for URLs.

Joe-Edwards-GDS reviewed (commented):
Thanks @sammyiyke , I think this is looking good. Also needs a merge/rebase with main

sammyiyke committed:
PYIC-5669 - Adds PYI_UPDATE_DETAILS to PAGES

sammyiyke committed:
PYIC-5669 - Rename pyi-update-details to update-details

Inline with dropping the pyi prefix for urls this commit drops the prefix for the new page

sammyiyke committed:
PYIC-5669 - Moved logic for nameParts

Moved logic for handling nameParts into the reuseHelper.js

- Removes logic from template
- Makes it more accessible for future use

sammyiyke committed:
PYIC-5669 - Lint fix

Joe-Edwards-GDS commented on src/app/ipv/middleware.test.js:
```diff
@@ -515,6 +519,10 @@ describe("journey middleware", () => {
       const expectedUserDetail = {
         name: "firstName LastName",
+        nameParts: {
+          givenName: "firstName",
```
Do you think this should test the case where you have multiple given names?

sammyiyke replied:
I think it might be better to add that test case as a new test

Joe-Edwards-GDS commented on src/app/shared/reuseHelper.js:
```diff
@@ -50,8 +54,23 @@ module.exports = {
+        } else if (namePart.type === "FamilyName") {
+          familyName = namePart.value;
+        }
+      });
```
I don't think this is right - what if you have multiple given names?

What about something like:
```
const givenName = userDetailsResponse.data?.nameParts
  .filter((namePart) => namePart.type === "GivenName")
  .join(" ");
```

sammyiyke replied:
Went with a slightly different approach, idea is to try reducing the operations. Instead of using two separate `filter`/`find` operations

Joe-Edwards-GDS commented on src/views/ipv/page/update-details.njk:
```diff
@@ -14,35 +14,21 @@
         },
         value: {
-          text: givenNames.join(' ')
+          text: userDetails.nameParts.givenName | safe
```
Why are we using `safe` here? I think we do want to escape this

sammyiyke committed:
PYIC-5669 - PR Review related fixes

- Removes unnecessary `safe` escape in template

- Handles multiple given names in reuseHelper.js

- Add new test for multiple given names

sammyiyke committed:
PYIC-5669 - PR Review related fixes

- Removes unnecessary `safe` escape in template

- Handles multiple given names in reuseHelper.js

- Add new test for multiple given names

sammyiyke committed:
PYIC-5669 - PR Review related fixes

- Handles multiple given last in reuseHelper.js

sammyiyke committed:
PYIC-5669 - PR Review related fixes

- Handles multiple given last in reuseHelper.js
